File: Java_serial/from_generator_example/DARNp_generator.java
import java.util.*;

class DARNp_generator {
	
	public DARNp_generator() {
		
	}
	
	//Generate Bernouli random variable.
	private static boolean bernoulli(Double p, Random rand) {
		if (rand.nextDouble() < p) {
			return true;
		}
		else {
			return false;
		}
	}
	
	//Generate DARNp edge series.
	public static boolean[][] generate_DARNp_net(int nodes, int mem_leng, double mem_str, double avg_prob, int time_steps) {
		
		Random rand = new Random();
		int i,t;
		int past_val;
		boolean mem_check;
		
		int num_links = (nodes * (nodes-1))/2;
		
		//Initialise array of bools for the edge series.
		boolean[][] edge_series = new boolean[num_links][time_steps+mem_leng];
		
		//Make edge series. Note that since links are independent we do this link by link.
		for (i=0 ; i<num_links ; i++) {
			//Initialise the memory of the network
			for (t=0 ; t<mem_leng ; t++) {
				edge_series[i][t] = bernoulli(avg_prob, rand);
			}
			
			//Now fill the edge list.
			for (t=mem_leng ; t<time_steps+mem_leng ; t++) {
				mem_check = bernoulli(mem_str, rand);
				
				if (mem_check) {
					past_val = rand.nextInt(mem_leng)+1;
					edge_series[i][t] = edge_series[i][t-past_val];
				}
				else {
					edge_series[i][t] = bernoulli(avg_prob, rand);
				}
			}
		}
		
		return edge_series;
	}
	
	public static void main(String[] args) {
		
	}
}

File: python_2/from_generator/DARNp_generator.py
#!/usr/bin/python
import random

"""
This program contains functions to generate a simple DARN(p) temporal network.
This network has memory precicely p.
"""


def main():
	
	print("Program build correctly!")
	print("Note, this program is not intended to run on its own...")
	print("This is the only output.")
	
	return


"""
Function to generate DARN(p) network.
Takes (in order):
	- Number of nodes.
	- Memory length.
	- Memory strength (reliance on memory).
	- Link probability.
	- Timesteps to run model for.
returns edge sequence in format required for memory estimator.
"""
def generate_DARNp_net(nodes, mem_leng, mem_str, avg_prob, time_steps):
	
	num_links = (nodes * (nodes-1))/2		#Number of links in this model.
	
	#Initialise edge sequence, add extra time steps for initialising memory.
	edge_sequence = [[0]*(time_steps+mem_leng)]*num_links
	
	#Initialise memory.
	for t in xrange(mem_leng):
		for l in xrange(num_links):
			edge_sequence[l][t] = bool(bernoulli(avg_prob))
		
	
	#Now run model.
	for t in range(mem_leng,time_steps):
		for l in xrange(num_links):
			mem_check = bernoulli(mem_str)
			
			if mem_check == 0:
				edge_sequence[l][t] = bool(bernoulli(avg_prob))
			else:
				past_val = random.randint(1,mem_leng)
				edge_sequence[l][t] = edge_sequence[l][t-past_val]
	
	return edge_sequence


"""
Function generates Bernoulli random variables with probability p.
"""
def bernoulli(p):

	r_num = random.random()
	if r_num < p:
		return 1
	else:
		return 0

if __name__ == "__main__":
	main()